broker.go

```diff
@@ -352,7 +352,7 @@ func (b *Broker) RemovePipeline(t EventType, id PipelineID) error {
 //
 // Once we start deleting nodes, we will continue until completion, but we'll
 // return false with an error.
-func (b *Broker) RemovePipelineAndNodes(t EventType, id PipelineID) (bool, error) {
+func (b *Broker) RemovePipelineAndNodes(ctx context.Context, t EventType, id PipelineID) (bool, error) {
```

Comment:
I don't think we should test for cancellation as a precondition or within the for loop that deletes nodes.  

Let's talk about use-cases... if RemovePipelineAndNodes is called with an API ctx, the request could timeout and be cancelled, which might leave the state of the graph in a bit of a mess.  

At the moment, the ctx is passed to the Close() op which can check for cancellation appropriately, but even if that fails the node will get deleted and leave the graph in a good state.

Reply:
I think my wording could have done with more finessing to explain my thoughts. 😞 

> I don't think we should test for cancellation as a precondition or within the for loop that deletes nodes.

I agree with this, I feel once we've started, the deletion of pipeline and nodes should be atomic.

I think what I meant by:

> any method which receives Context?

... was actually more like: any method that crosses the domain boundary into the eventlogger (i.e. `RemovePipelineAndNodes` could be checked before a long running or delicate operation was begun. Sorry I didn't do a great job of explaining earlier. 

I assume passing context to `Close` is mostly about allowing the node the opportunity to get things from the context that it might need to close successfully?

Reply:
Passing ctx to Close() serves two purposes: 

1. Allows it to dig out ctx specific data
2. Allows it to pass ctx to other operations.  For instance, it's very likely that a Close() implementation might make a call back to the broker to do something like process a msg, and it would want to pass along to the ctx so the broker can do things like check for cancellation.